### PR TITLE
feat: `avm/res/container-service/managed-cluster` Updates container service to support automatic clusters

### PR DESCRIPTION
**Pull request by zedy-wj**
## Description

Related to PR: https://github.com/Azure/azure-dev/pull/4069

Adds new automatic setting for container service managed cluster.

Set `azure/login@v2` to use federated credential. (Due to some security policies, we are currently unable to use client secret for azure login. After completing all changes to this PR, we will restore the login method. This is currently only for testing the github pipeline.)

The failed test in the Pipeline Status Badge has nothing to do with this PR. The specific information can be viewed at: https://github.com/zedy-wj/bicep-registry-modules/actions/runs/10053355723/job/27786049197#step:4:649.

## Pipeline Reference

| Pipeline |
| -------- |
| [avm.res.container-service.managed-cluster](https://github.com/zedy-wj/bicep-registry-modules/actions/workflows/avm.res.container-service.managed-cluster.yml)         |

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

@jongio for notification.

### CODE REVIEW

File: .github/actions/templates/avm-validateModuleDeployment/action.yml [hunk 1]
```diff
@@ -62,7 +62,9 @@ runs:
     - name: Azure Login
       uses: azure/login@v2
       with:
-        creds: ${{ env.AZURE_CREDENTIALS }}
+        client-id: ${{ env.AZURE_CLIENT_ID }}
```
Comment (AlexanderSehr) [inline .github/actions/templates/avm-validateModuleDeployment/action.yml:65]: Unrelated changes to this and the `avm.template.module.yml` file

File: avm/res/container-service/managed-cluster/main.bicep [hunk 2]
```diff
@@ -753,6 +759,15 @@ resource managedCluster 'Microsoft.ContainerService/managedClusters@2024-03-02-p
   }
 }
 
+module managedCluster_maintenanceConfigurations 'maintenance-configurations/main.bicep' = if (automatic) {
+  name: '${uniqueString(deployment().name, location)}-ManagedCluster-MaintenanceConfigurations'
+  params: {
+    name: 'aksManagedAutoUpgradeSchedule'
```
Comment (AlexanderSehr) [inline avm/res/container-service/managed-cluster/main.bicep:765]: Shoud this be hardcoded (e.g., because it must be a certain value). If so, I'd recommend to move it to the child as a default and remove the param here.

File: avm/res/container-service/managed-cluster/main.bicep [hunk 3]
```diff
@@ -156,6 +156,12 @@ param primaryAgentPoolProfile array
 @description('Optional. Define one or more secondary/additional agent pools.')
 param agentPools agentPoolType
 
+@description('Optional. Whether or not to use AKS Automatic mode.')
+param automatic bool = false
+
+@description('Optional. Maintenance window for the maintenance configuration.')
+param maintenanceWindow object = {}
```
Comment (AlexanderSehr) [inline avm/res/container-service/managed-cluster/main.bicep:163]: Instead of 2 parameters, it may make sense to convert this to a single `param maintenanceConfiguration object?` parameter (ideally using a UDT), and update the deployment below to have a condition like `module managedCluster_maintenanceConfigurations 'maintenance-configurations/main.bicep' = if (!empty(maintenanceConfiguration)) { (...) }`
This way you can dynamically update it later without the need of additional parameters. Also, if using a UDT you could then declare the `maintenanceWindow` parameter as mandatory IF the `maintainanceConfiguartion` would be defined.

Comment (jongio) [reply avm/res/container-service/managed-cluster/main.bicep:163]: UDTs are required for params right?

File: .github/actions/templates/avm-validateModuleDeployment/action.yml (hunk 1 above)
Comment (zedy-wj) [reply .github/actions/templates/avm-validateModuleDeployment/action.yml:65]: Yes, this is for testing purposes only. After all changes are complete, we will revert the changes to these two files. 😃

**Comment by zedy-wj** [comment]
Hey, @AlexanderSehr  - Updated according to your comments. Please re-review it, thanks!

File: avm/res/container-service/managed-cluster/main.bicep (hunk 3 above)
Comment (AlexanderSehr) [reply avm/res/container-service/managed-cluster/main.bicep]: Required no, recommended yes, absolutely 🚀

**Comment by AlexanderSehr** [comment]
Hey @ilhaan & @JPEasier, I took the liberty to add some comments that @zedy-wj already addressed. 
As you're the owners, please take a look when you get the chance and comment/approve as you see fit 💪

**Comment by ilhaan** [comment]
This looks good to me but I'm concerned about the failing pipeline. @JPEasier are you aware of the pipeline failure error?

**Comment by zedy-wj** [comment]
> ently unable to use client secret for azure login. After completing all changes to this PR, we will restore the login method. This is currently only for testing the github

It seems that because of the resource quota under the subscription, after I changed the region, the pipeline can pass normally.

**Comment by zedy-wj** [comment]
@AlexanderSehr, @ilhaan, @JPEasier - Since the pipeline status badge has passed and you have reviewed it, I removed the changes that were not related to this PR (login with fed credential). If you don't have any other questions, could you help me merge this PR, thanks!

**Comment by AlexanderSehr** [comment]
Looks good to me, but I'll leave it to the module owners to give their blessing and approve the PR :)

**Review by ilhaan — APPROVED** [review]
@zedy-wj Approved. Thank you!
